pkg/forwarders/tlsforwarder_test.go

```diff
@@ -75,126 +137,72 @@ f3cb9gYaLWdmvkx8p3g=
+			Certificates:       []tls.Certificate{tlsCert},
 			InsecureSkipVerify: true, //nolint:gosec
 		},
 	)
 	require.NoError(t, err)
-	require.NotNil(t, listener)
 	t.Logf("TLS server address %s", listener.Addr().String())
 
-	var wg sync.WaitGroup
-	go acceptLoop(t, listener, &wg,
-		[]string{
-			"<14>signal=test-signal;key=value;\n",
-			"<14>signal=test-signal-2;key=value;\n",
-		},
-	)
-
-	serializer := serializers.NewSemicolonDelimited()
-	require.NotNil(t, serializer)
-
-	log := logrusr.New(logrus.New())
-	tf, err := NewTLSForwarder(listener.Addr().String(), log,
-		func(c *tls.Config) {
-			c.RootCAs = roots
-			c.Certificates = []tls.Certificate{tlsCert}
-			c.InsecureSkipVerify = true
-		},
-	)
-	require.NoError(t, err)
-	require.NotNil(t, tf)
-
-	consumer := telemetry.NewConsumer(serializer, tf)
-
-	m, err := telemetry.NewManager(
-		"test-ping",
-		telemetry.OptManagerPeriod(time.Hour),
-		telemetry.OptManagerLogger(log),
-	)
-	require.NoError(t, err)
-
-	w := telemetry.NewWorkflow("test1")
-	p, err := provider.NewFixedValueProvider("test1-provider", types.ProviderReport{
-		"key": "value",
-	})
-	require.NoError(t, err)
-	w.AddProvider(p)
-	m.AddWorkflow(w)
-	require.NoError(t, m.AddConsumer(consumer))
-	require.NoError(t, m.Start())
+	return telemetryServer{
+		listener: listener,
+		done:     make(chan struct{}),
+	}
+}
 
-	wg.Add(1)
-	require.NoError(t, m.TriggerExecute(context.Background(), "test-signal"))
-	wg.Add(1)
-	require.NoError(t, m.TriggerExecute(context.Background(), "test-signal-2"))
-	wg.Wait()
+func (ts telemetryServer) Addr() string {
+	return ts.listener.Addr().String()
+}
 
-	require.NoError(t, listener.Close())
-	m.Stop()
+func (ts telemetryServer) Done() {
+	<-ts.done
 }
 
-func acceptLoop(t *testing.T, l net.Listener, wg *sync.WaitGroup, expectedData []string) {
+func (ts telemetryServer) RunAndAssertExpectedData(ctx context.Context, t *testing.T, expectedData []string) {
 	t.Log("server: accepting...")
-
-	// Accept just once because TLSForwarder persists the connection
-	conn, err := l.Accept()
-	if err != nil {
-		t.Logf("server: Accept() returned error: %v", err)
-		return
+	receivedData := make(chan string, len(expectedData))
+	go func() {
+		for {
+			conn, err := ts.listener.Accept()
+			if err != nil {
+				t.Logf("server: Accept() returned error: %v", err)
+				return
+			}
+			go handleConnection(t, conn, receivedData)
+		}
+	}()
+	defer ts.listener.Close()
+	for _, expected := range expectedData {
+		select {
+		case received := <-receivedData:
+			assert.Equal(t, expected, received)
+		case <-ctx.Done():
+			assert.Failf(t, "timeout waiting for data", "expected data: %q", expected)
+		}
 	}
-	handleClient(t, conn, expectedData, wg)
+	ts.done <- struct{}{}
 }
 
-func handleClient(t *testing.T, conn net.Conn, expectedData []string, wg *sync.WaitGroup) {
+func handleConnection(t *testing.T, conn net.Conn, receivedData chan<- string) {
```

Comment:
Do you mind adding a godoc here while we're at it? I reckon this deserves a small explanation what it does (it already did when we first wrote it 😅 )